File: Code Samples/Linked List - Harsh Duche/PYTHON/Circular.py
class CircularNode:
    def __init__(self, data):
        self.data = data
        self.next = None

class CircularLinkedList:
    def __init__(self):
        self.head = None

    # Create operation: Adding a new node to the end
    def add_node(self, data):
        new_node = CircularNode(data)
        if self.head is None:
            self.head = new_node
            new_node.next = self.head
        else:
            current = self.head
            while current.next != self.head:
                current = current.next
            current.next = new_node
            new_node.next = self.head

    # Read operation: Accessing elements by index
    def get_node(self, index):
        if self.head is None:
            return -1  # List is empty
        current = self.head
        for i in range(index):
            current = current.next
            if current == self.head:
                return -1  # Index out of bounds
        return current.data

    # Update operation: Modifying an element by index
    def update_node(self, index, data):
        if self.head is None:
            return  # List is empty
        current = self.head
        for i in range(index):
            current = current.next
            if current == self.head:
                return  # Index out of bounds
        current.data = data

    # Delete operation: Removing a node by index (without additional data structures)
    def delete_node(self, index):
        if self.head is None:
            return  # List is empty
        current = self.head
        prev = None
        for i in range(index):
            prev = current
            current = current.next
            if current == self.head:
                return  # Index out of bounds
        if current == self.head:
            # Removing the head node
            if self.head.next == self.head:
                self.head = None
            else:
                prev = self.head
                while prev.next != self.head:
                    prev = prev.next
                self.head = self.head.next
                prev.next = self.head
        else:
            prev.next = current.next

    # isEmpty operation
    def is_empty(self):
        return self.head is None

# Example usage
circular_linked_list = CircularLinkedList()

# Create operation
circular_linked_list.add_node(1)
circular_linked_list.add_node(2)
circular_linked_list.add_node(3)

# Read operation
print("Element at index 1:", circular_linked_list.get_node(1))

# Update operation
circular_linked_list.update_node(1, 4)

# Delete operation
circular_linked_list.delete_node(1)

# isEmpty operation
print("Is the list empty?", circular_linked_list.is_empty())

File: Code Samples/Linked List - Harsh Duche/CPP/Circular.cpp
#include <iostream>

class Node {
public:
    int data;
    Node* next;

    Node(int data) : data(data), next(nullptr) {}
};

class CircularLinkedList {
public:
    Node* head;

    CircularLinkedList() : head(nullptr) {}

    // Create operation: Adding a new node to the end
    void addNode(int data) {
        Node* newNode = new Node(data);
        if (head == nullptr) {
            head = newNode;
            newNode->next = head;
        } else {
            Node* current = head;
            while (current->next != head) {
                current = current->next;
            }
            current->next = newNode;
            newNode->next = head;
        }
    }

    // Read operation: Accessing elements by index
    int getNode(int index) {
        if (head == nullptr) {
            return -1; // List is empty
        }
        Node* current = head;
        for (int i = 0; i < index; i++) {
            current = current->next;
            if (current == head) {
                return -1; // Index out of bounds
            }
        }
        return current->data;
    }

    // Update operation: Modifying an element by index
    void updateNode(int index, int data) {
        if (head == nullptr) {
            return; // List is empty
        }
        Node* current = head;
        for (int i = 0; i < index; i++) {
            current = current->next;
            if (current == head) {
                return; // Index out of bounds
            }
        }
        current->data = data;
    }

    // Delete operation: Removing a node by index (without additional data structures)
    void deleteNode(int index) {
        if (head == nullptr) {
            return; // List is empty
        }
        Node* current = head;
        Node* prev = nullptr;
        for (int i = 0; i < index; i++) {
            prev = current;
            current = current->next;
            if (current == head) {
                return; // Index out of bounds
            }
        }
        if (current == head) {
            // Removing the head node
            if (head->next == head) {
                head = nullptr;
            } else {
                prev = head;
                while (prev->next != head) {
                    prev = prev->next;
                }
                head = head->next;
                prev->next = head;
            }
        } else {
            prev->next = current->next;
        }
        delete current;
    }

    // isEmpty operation
    bool isEmpty() {
        return head == nullptr;
    }
};

int main() {
    CircularLinkedList list;

    // Create operation
    list.addNode(1);
    list.addNode(2);
    list.addNode(3);

    // Read operation
    std::cout << "Element at index 1: " << list.getNode(1) << std::endl;

    // Update operation
    list.updateNode(1, 4);

    // Delete operation
    list.deleteNode(1);

    // isEmpty operation
    std::cout << "Is the list empty? " << list.isEmpty() << std::endl;

    return 0;
}
